// next-alert/dist/components/Alert.js
"use strict";
"use client";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/Alert.tsx
var Alert_exports = {};
__export(Alert_exports, {
  Alert: () => Alert
});
module.exports = __toCommonJS(Alert_exports);
var import_react2 = require("react");

// src/components/AlertContext.tsx
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
var AlertContext = (0, import_react.createContext)(void 0);
var useAlert = () => {
  const context = (0, import_react.useContext)(AlertContext);
  if (!context) {
    throw new Error("useAlert must be used within an AlertProvider");
  }
  return context;
};

// src/components/Alert.tsx
var import_framer_motion = require("framer-motion");
var import_jsx_runtime2 = require("react/jsx-runtime");
var Alert = ({
  idElement,
  message,
  idDelay,
  type = "success",
  duration = 3e3,
  title,
  children,
  passFunction,
  direction = "right",
  ...restProps
}) => {
  const { removeAlert } = useAlert();
  (0, import_react2.useEffect)(() => {
    setTimeout(
      () => {
        removeAlert(idElement);
        if (passFunction) {
          passFunction();
        }
      },
      // set delay if user click to fast
      duration + idDelay * 500
    );
  }, [idElement, removeAlert]);
  const colorOfAlert = type === "success" && { backgroundColor: "#16a34a" } || type === "error" && { backgroundColor: "#ef4444" } || type === "warning" && { backgroundColor: "#facc15" } || type === "info" && { backgroundColor: "#0ea5e9" };
  const directionOfAlert = direction === "left" && { x: -100 } || direction === "right" && { x: 100 } || direction === "top" && { y: -100 } || direction === "bottom" && { y: 100 };
  const directionOfAlertXY = direction === "left" && { x: 0 } || direction === "right" && { x: 0 } || direction === "top" && { y: 0 } || direction === "bottom" && { y: 0 };
  return /* @__PURE__ */ (0, import_jsx_runtime2.jsxs)(
    import_framer_motion.motion.div,
    {
      ...restProps,
      initial: { ...directionOfAlert, opacity: 0 },
      animate: { ...directionOfAlertXY, opacity: 1 },
      exit: { ...directionOfAlert, opacity: 0 },
      style: {
        overflow: "hidden",
        ...colorOfAlert,
        color: "white",
        width: "15rem",
        padding: "1rem",
        margin: "0.5rem",
        minWidth: "fit-content"
      },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime2.jsx)("div", { style: { "position": "absolute", "top": "0", "left": "0", "width": "100%" }, children: /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
          import_framer_motion.motion.div,
          {
            initial: { width: 0 },
            animate: { width: "100%" },
            transition: { duration: duration / 1e3 },
            style: {
              height: "0.3rem",
              opacity: 0.5,
              backgroundColor: "white"
            }
          }
        ) }),
        children,
        /* @__PURE__ */ (0, import_jsx_runtime2.jsx)("h1", { children: /* @__PURE__ */ (0, import_jsx_runtime2.jsx)("b", { children: title }) }),
        /* @__PURE__ */ (0, import_jsx_runtime2.jsx)("p", { children: message })
      ]
    }
  );
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Alert
});
