// next-alert/dist/components/Alert.mjs
"use client";

// src/components/Alert.tsx
import { useEffect } from "react";

// src/components/AlertContext.tsx
import { createContext, useContext, useState } from "react";
import { jsx } from "react/jsx-runtime";
var AlertContext = createContext(void 0);
var useAlert = () => {
  const context = useContext(AlertContext);
  if (!context) {
    throw new Error("useAlert must be used within an AlertProvider");
  }
  return context;
};

// src/components/Alert.tsx
import { motion } from "framer-motion";
import { jsx as jsx2, jsxs } from "react/jsx-runtime";
var Alert = ({
  idElement,
  message,
  idDelay,
  type = "success",
  duration = 3e3,
  title,
  children,
  passFunction,
  direction = "right",
  ...restProps
}) => {
  const { removeAlert } = useAlert();
  useEffect(() => {
    setTimeout(
      () => {
        removeAlert(idElement);
        if (passFunction) {
          passFunction();
        }
      },
      // set delay if user click to fast
      duration + idDelay * 500
    );
  }, [idElement, removeAlert]);
  const colorOfAlert = type === "success" && { backgroundColor: "#16a34a" } || type === "error" && { backgroundColor: "#ef4444" } || type === "warning" && { backgroundColor: "#facc15" } || type === "info" && { backgroundColor: "#0ea5e9" };
  const directionOfAlert = direction === "left" && { x: -100 } || direction === "right" && { x: 100 } || direction === "top" && { y: -100 } || direction === "bottom" && { y: 100 };
  const directionOfAlertXY = direction === "left" && { x: 0 } || direction === "right" && { x: 0 } || direction === "top" && { y: 0 } || direction === "bottom" && { y: 0 };
  return /* @__PURE__ */ jsxs(
    motion.div,
    {
      ...restProps,
      initial: { ...directionOfAlert, opacity: 0 },
      animate: { ...directionOfAlertXY, opacity: 1 },
      exit: { ...directionOfAlert, opacity: 0 },
      style: {
        overflow: "hidden",
        ...colorOfAlert,
        color: "white",
        width: "15rem",
        padding: "1rem",
        margin: "0.5rem",
        minWidth: "fit-content"
      },
      children: [
        /* @__PURE__ */ jsx2("div", { style: { "position": "absolute", "top": "0", "left": "0", "width": "100%" }, children: /* @__PURE__ */ jsx2(
          motion.div,
          {
            initial: { width: 0 },
            animate: { width: "100%" },
            transition: { duration: duration / 1e3 },
            style: {
              height: "0.3rem",
              opacity: 0.5,
              backgroundColor: "white"
            }
          }
        ) }),
        children,
        /* @__PURE__ */ jsx2("h1", { children: /* @__PURE__ */ jsx2("b", { children: title }) }),
        /* @__PURE__ */ jsx2("p", { children: message })
      ]
    }
  );
};
export {
  Alert
};
